### PR TITLE
James 1901 Bring metrics into ES

### PR DESCRIPTION
**Pull request by chibenwa**
Evolution of : https://github.com/linagora/james-project/pull/547

Integration test added. Extracted to a new project.

Also ES has its backend project. Better isolation of components and responsability with ES.

### CODE REVIEW

**Comment by chibenwa** [comment]
```
[Building Java 8 james-1901-remix-3def1c5e-d4c3-47c5-b115-c69fe78679fa]   CassandraVacationIntegrationTest>VacationIntegrationTest.jmapVacationShouldSendNotificationTwiceWhenVacationReset:267 » ConditionTimeout
```

File: server/container/metrics/metrics-dropwizard/src/main/java/org/apache/james/metrics/dropwizard/ESMetricReporter.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+            } catch (IOException e) {
+                throw Throwables.propagate(e);
+            }
+        }   
+        return Optional.absent();
+    }
+
+    public void start() {
+        if (reporter.isPresent()) {
+            reporter.get().start(esReporterConfiguration.getPeriodInSecond(), TimeUnit.SECONDS);
```
Comment (rouazana) [inline server/container/metrics/metrics-dropwizard/src/main/java/org/apache/james/metrics/dropwizard/ESMetricReporter.java]: reporter.ifPresent(...)?

File: server/container/metrics/metrics-dropwizard/src/main/java/org/apache/james/metrics/dropwizard/ESMetricReporter.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        }
+    }
+
+    @PreDestroy
+    public void stop() {
+        if (reporter.isPresent()) {
+            reporter.get().stop();
```
Comment (rouazana) [inline server/container/metrics/metrics-dropwizard/src/main/java/org/apache/james/metrics/dropwizard/ESMetricReporter.java]: reporter.ifPresent(reporter::stop)?

File: server/container/metrics/metrics-dropwizard/src/main/java/org/apache/james/metrics/dropwizard/ESReporterConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    }
+
+    public static ESReporterConfiguration enabled(String host, int port, Optional<String> index, Optional<Long> periodInSecond) {
+        return new ESReporterConfiguration(
+            Optional.of(host),
+            Optional.of(port),
+            ENABLED,
+            index,
+            periodInSecond);
+    }
```
Comment (rouazana) [inline server/container/metrics/metrics-dropwizard/src/main/java/org/apache/james/metrics/dropwizard/ESReporterConfiguration.java]: can you make a builder? There are too much basic misleading types in this constructor (int/long and 2 strings)

File: server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/ElasticSearchMailboxModule.java [hunk 4]
```diff
@@ -50,6 +50,9 @@
     private static final int DEFAULT_CONNECTION_MAX_RETRIES = 7;
     private static final int DEFAULT_CONNECTION_MIN_DELAY = 3000;
     private static final boolean DEFAULT_INDEX_ATTACHMENTS = true;
+    public static final String ES_CONFIG_FILE = FileSystem.FILE_PROTOCOL_AND_CONF + "elasticsearch.properties";
+    public static final String ELASTICSEARCH_MASTER_HOST = "elasticsearch.masterHost";
+    public static final String ELASTICSEARCH_PORT = "elasticsearch.port";
```
Comment (rouazana) [inline server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/ElasticSearchMailboxModule.java]: public before private?

File: backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/DeleteByQueryPerformer.java [hunk 5]
```diff
@@ -43,17 +43,21 @@
     private final Client client;
     private final ExecutorService executor;
     private final int batchSize;
+    private final String indexName;
+    private final String typeName;
 
     @Inject
-    public DeleteByQueryPerformer(Client client, @Named("AsyncExecutor") ExecutorService executor) {
-        this(client, executor, DEFAULT_BATCH_SIZE);
+    public DeleteByQueryPerformer(Client client, @Named("AsyncExecutor") ExecutorService executor, String indexName, String typeName) {
```
Comment (rouazana) [inline backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/DeleteByQueryPerformer.java]: Can you define custom type to differentiate indexName and typeName by strong typing?

File: server/container/metrics/metrics-dropwizard/src/main/java/org/apache/james/metrics/dropwizard/ESMetricReporter.java (hunk 1 above)
Comment (chibenwa) [inline server/container/metrics/metrics-dropwizard/src/main/java/org/apache/james/metrics/dropwizard/ESMetricReporter.java]: Already done ;-)

File: server/container/metrics/metrics-dropwizard/src/main/java/org/apache/james/metrics/dropwizard/ESMetricReporter.java (hunk 2 above)
Comment (chibenwa) [inline server/container/metrics/metrics-dropwizard/src/main/java/org/apache/james/metrics/dropwizard/ESMetricReporter.java]: Already done ;-)

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**